### ISSUE TITLE
decide the optimal count of total vnode

### ISSUE THREAD

**Issue by BugenZhao** [risingwavelabs/risingwave#3579]
We decreased vnode count from 2048 to 256 in #3558 since the performance is poor to create 2048 iterators and merge-sort them in a single batch table scan. This could be improved by

- Decide whether to guarantee order for batch table scan, so that we can decrease the number of iterators and usages of merge-sort as much as possible.
- Support real parallel scan, so there's no need to scan all vnodes in a single parallel unit.
- Optimize iterator creation in Hummock.

**Comment by hzxa21** [comment]
per our discussion with @fuyufjh and @xx01cyx , i think the plan is to set the number of vnode used in a mv to be 4 * （number_of_parallel_unit when the mv is created）

**Comment by jon-chuang** [comment]
@hzxa21 but in the event of scale up beyond 4x, state migration cost could be quite bad, right? Is there a way to incrementally migrate to a new vnode mapping?

**Comment by fuyufjh** [comment]
> but in the event of scale up beyond 4x, state migration cost could be quite bad, right? Is there a way to incrementally migrate to a new vnode mapping?

Yep. In this design, migrating to higher parallelism must need a full data rewriting.

> So, when encountering SST with old vnode count, one can just scan the vnode prefix (originally 4 bits, now 5 bits, so scan the first 4 bits, then return only those vnodes with 5th bit the desired bit). So the scan cost is doubled. Then those SSTs that have been marked as migrated can just scan the first 5 bits.

This might not work. The storage must decide the ordering of keys when persisting them into SSTs. Since we decided to put vnode as a prefix of primary key in storage, the (full) vnode number will decide the ordering.

**Comment by jon-chuang** [comment]
> This might not work. Since we decided to put vnode as a prefix of primary key in storage, the (full) vnode number will decide the ordering when persisting KV entries into SSTs.

Hmm, yes. So maybe, for migrated: use 5 bits for vnode, and prefix scan first 4 bits, for unmigrated: use first 4 bits as usual. Then when fully migrated (all old SSTs have been rewritten) then we can use all the 5 bits.

**Comment by BugenZhao** [comment]
Superseded by #15900.

**Comment by BugenZhao** [comment]
Since #15900 has been done, we now gain the capability for assigning different vnode count for different jobs.

Specifying a maximum parallelism greater than 256 might be uncommon for users (except for the ones with heavy workloads). However, we may consider setting a lower default maximum parallelism (like 64), which may enhance performance for most users.

Thus, we may revisit this issue to select a more optimal vnode count based on benchmarking results.
